### DISCUSSION TITLE
OpenTelemetry vs Prometheus label names

### DISCUSSION THREAD

**Discussion by arendjr** [autometrics-dev/autometrics-shared#93] [General]
I noticed the [draft spec](https://github.com/autometrics-dev/autometrics-shared/pull/85) contains both OpenTelemetry and Prometheus specifications for all metric names. Most notably, OpenTelemetry metric names use `.` as a separator, while Prometheus metric names use `_`. The Prometheus ones also sometimes have suffixes added.

The reason for this dual-naming is because Prometheus conventions are codified in the [OpenMetrics standard]([https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#abnf)). For anyone curious to dive into the rabbit hole, we've had a previous OpenTelemetry vs OpenMetrics discussion on this: https://github.com/orgs/autometrics-dev/discussions/37

Unfortunately, we haven't considered this situation with regards to labels yet. Labels use `.` as separator in the draft spec, which is great according to OpenTelemetry, but is **forbidden** according to the [OpenMetrics ABNF](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#abnf). Our current clients also use underscores as separators AFAIK, so they're de-facto following OpenMetrics today.

Fortunately, my recommendation would be simple: We keep the Autometrics draft spec as is, but suggest that all dots (`.`) in label names should be replaced with underscores (`_`) when submitting to Prometheus.

**Comment by brettimus** [comment]
seems raisonable

**Comment by hatchan** [comment]
When you say "submitting to Prometheus", as you referring to the exposition format at the `/metrics` endpoints (and possibly the non otel push metrics to gravel/pushgateway)?

**Reply by arendjr** [reply]
Yes, exactly.

**Comment by hatchan** [comment]
How will the OTEL push flow be impacted by this? Will these payloads still contain the `.`?

**Reply by arendjr** [reply]
That's what I would expect, yes.

**Reply by hatchan** [reply]
I wonder what happens when you push a OTEL payload to a otel collector or directly to Prometheus ... Will it be translated :thinking: Maybe something that I can prototype tomorrow...

**Reply by gagbo** [reply]
I did this already, and the otel collector accepts the `.` without issue. I assume that the collector does the `_` translation when sending to Prometheus

**Reply by arendjr** [reply]
I noticed while implementing this PR that at least the TypeScript OTEL libraries convert the dots back to underscores for the Prometheus exporter for us: https://github.com/autometrics-dev/autometrics-ts/pull/140

So I think it's reasonable to expect OTLP => Prometheus conversions elsewhere to do this too.

**Reply by hatchan** [reply]
I can confirm what @gagbo said. The following flow works as expected (ie. no dots, just underscores): application -> otel collector -> remotewrite endpoint (prometheus).

**Comment by hatchan** [comment]
Would it maybe make sense to just avoid this whole issue, by not using the `.` altogether and use `_` for everything? We do use some of the same names that are present in the OTEL spec, but I think these are quite loosly coupled.

**Reply by arendjr** [reply]
My understanding was that would go against the Otel conventions/recommentations. For instance, Otel explicitly uses `service.name`, whereas in Prometheus it would become `service_name`.

**Comment by hatchan** [comment]
After looking through the draft 1.0 spec I found that this is already part of it: https://github.com/autometrics-dev/autometrics-shared/blob/7409f0225c5648906213e9401ae39ab17008b11d/specs/autometrics_v1.0.0.md?plain=1#L75-L77
